### ISSUE TITLE
Bug: typescript eslint raises warnings when used with react-hooks

### ISSUE THREAD

**Issue by HatemMn** [typescript-eslint/typescript-eslint#9126]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

ast-spec

### Repro Code

The reference to reproduce this issue is the code in the first message of this issue : https://github.com/facebook/react/issues/28313

If we use the same code propesed, ie : 

```TypeScript
// eslint.config.js
import eslint from "@eslint/js";
import hooksPlugin from "eslint-plugin-react-hooks";

export default [
  eslint.configs.recommended,
  {
    plugins: {
      "react-hooks": hooksPlugin,
    },
    rules: hooksPlugin.configs.recommended.rules,
  },
];
```

It works, however, if I do :

```TypeScript
import eslint from "@eslint/js";
import hooksPlugin from "eslint-plugin-react-hooks";
import tseslint from "typescript-eslint";

export default [
  eslint.configs.recommended,
  {
    plugins: {
      "react-hooks": hooksPlugin,
    },
    rules: hooksPlugin.configs.recommended.rules,
    ...tseslint.configs.recommendedTypeChecked,

  },
];
```

I get this error : 

```
Type '{ "react-hooks/rules-of-hooks": string; "react-hooks/exhaustive-deps": string; } | { 'react-hooks/rules-of-hooks': string; 'react-hooks/exhaustive-deps': string; }' is not assignable to type 'Partial<Record<string, RuleEntry>> | undefined'.
  Type '{ "react-hooks/rules-of-hooks": string; "react-hooks/exhaustive-deps": string; }' is not assignable to type 'Partial<Record<string, RuleEntry>>'.
    Property '"react-hooks/rules-of-hooks"' is incompatible with index signature.
      Type 'string' is not assignable to type 'RuleEntry | undefined'.ts(2322)
Config.d.ts(255, 9): The expected type comes from property 'rules' which is declared here on type 'ConfigWithExtends'
```

### ESLint Config

```js
import eslint from "@eslint/js";
import hooksPlugin from "eslint-plugin-react-hooks";
import tseslint from "typescript-eslint";

export default [
  eslint.configs.recommended,
  {
    plugins: {
      "react-hooks": hooksPlugin,
    },
    rules: hooksPlugin.configs.recommended.rules,
    ...tseslint.configs.recommendedTypeChecked,

  },
];
```

### tsconfig

```jsonc
{
  "extends": "./tsconfig.base.json",
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "moduleResolution": "Node16",
    "noEmit": true
  },
  "files": ["eslint.config.mjs", "knip.config.ts"]
}

with : 

{
  "compilerOptions": {
    "jsx": "react-jsx",
    "strict": true,
    "alwaysStrict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "allowSyntheticDefaultImports": true,
    "verbatimModuleSyntax": true,
    "noFallthroughCasesInSwitch": true,
    "forceConsistentCasingInFileNames": true
  }
}
```

### Expected Result

No error

### Actual Result

```
Type '{ "react-hooks/rules-of-hooks": string; "react-hooks/exhaustive-deps": string; } | { 'react-hooks/rules-of-hooks': string; 'react-hooks/exhaustive-deps': string; }' is not assignable to type 'Partial<Record<string, RuleEntry>> | undefined'.
  Type '{ "react-hooks/rules-of-hooks": string; "react-hooks/exhaustive-deps": string; }' is not assignable to type 'Partial<Record<string, RuleEntry>>'.
    Property '"react-hooks/rules-of-hooks"' is incompatible with index signature.
      Type 'string' is not assignable to type 'RuleEntry | undefined'.ts(2322)
Config.d.ts(255, 9): The expected type comes from property 'rules' which is declared here on type 'ConfigWithExtends'
```

### Additional Info

If I omit the ts complaints, there is no problems (     // @ts-ignore )

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `X.Y.Z` |
| `@typescript-eslint/parser`            | `X.Y.Z` |
| `@typescript-eslint/rule-tester`       | `X.Y.Z` |
| `@typescript-eslint/scope-manager`     | `X.Y.Z` |
| `@typescript-eslint/typescript-estree` | `X.Y.Z` |
| `@typescript-eslint/type-utils`        | `X.Y.Z` |
| `@typescript-eslint/utils`             | `X.Y.Z` |
| `TypeScript`                           | `X.Y.Z` |
| `ESLint`                               | `X.Y.Z` |
| `node`                                 | `X.Y.Z` |

**Comment by auvred** [comment]
> It works, however, if I do :
> 
> ```ts
> import eslint from "@eslint/js";
> import hooksPlugin from "eslint-plugin-react-hooks";
> import tseslint from "typescript-eslint";
> 
> export default [
>   eslint.configs.recommended,
>   {
>     plugins: {
>       "react-hooks": hooksPlugin,
>     },
>     rules: hooksPlugin.configs.recommended.rules,
>     ...tseslint.configs.recommendedTypeChecked,
> 
>   },
> ];
> ```

This code sample does not contain anything that could cause the error you mentioned. Could you please provide a minimal isolated reproduction so we can figure out if this is a problem on the typescript-eslint side or on the `eslint-plugin-react-hooks` side?

**Comment by HatemMn** [comment]
I have a question before trying that : did you not have the error I am talking about despite using my same tsconfig ? here is it without the useless stuff : 

```

  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "moduleResolution": "Node16",
    "noEmit": true,
    "jsx": "react-jsx",
    "strict": true,
    "alwaysStrict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "allowSyntheticDefaultImports": true,
    "verbatimModuleSyntax": true,
    "noFallthroughCasesInSwitch": true,
    "forceConsistentCasingInFileNames": true
  },
  "files": ["eslint.config.mjs"]
```

I also have `// @ts-check` on top of my eslint file

**Comment by auvred** [comment]
The default exported array in the code example you provided is an untyped array. There are no places where the  `Type ... is not assignable to type ... .ts(2322)` error could occur here.

**Comment by bradzacher** [comment]
Youre trying to manually extract the rules from the plugin's legacy config. Because the module is untyped TS attempts to infer a type from the JS source code - hence it infers the crazy object type.

TS doesn't strictly infer the type - so it ignores the literal value types and generalises the object value as `string` - hence the failure.

You should not be manually pulling apart configs like this. You're doing flat configs wrong!

Instead use the compatibility tooling: https://www.npmjs.com/package/@eslint/eslintrc
This will let you convert the plugin's legacy configs to flat configs:
```ts
import { FlatCompat } from '@eslint/eslintrc';
import reactHooksPlugin from 'eslint-plugin-react-hooks';
 import tseslint from 'typescript-eslint';

const compat = new FlatCompat({ baseDirectory: import.meta.dirname });

export default tseslint.config([
  ... compat.config(reactHooksPlugin.configs.recommended),
  ...tseslint.configs.recommendedTypeChecked,
]);
```

**Comment by HatemMn** [comment]
Thank you for your reply @bradzacher  !
However, when I use the compatibility utility as you suggest, I now have the following problems reported by eslint : `Unsafe assignment of an `any` value.eslint[@typescript-eslint/no-unsafe-assignment](https://typescript-eslint.io/rules/no-unsafe-assignment)` and `Unsafe spread of an `any` value in an array.eslint[@typescript-eslint/no-unsafe-assignment](https://typescript-eslint.io/rules/no-unsafe-assignment)`

Is it normal ? Or I should not be linting my lintconfig ?

**Comment by HatemMn** [comment]
thank you for your amazing help @bradzacher  !

I know understand better, I changed my configs, they work, however I have one new problem and if you would please tell me why, typescript eslint complains about its own configuration ? ie : 
```
const compat = new FlatCompat({ baseDirectory: import.meta.dirname });
```
gives : 
```
   7:7   error  Unsafe assignment of an `any` value              @typescript-eslint/no-unsafe-assignment
```
Among other errors

I ignored my linting config for now, but is this normal ?

**Comment by bradzacher** [comment]
Yes - because you don't have types for `@eslint/eslintrc` (https://www.npmjs.com/package/@types/eslint__eslintrc).
No types means TS types the import as `any` - hence the rule errors.
